### DISCUSSION TITLE
Module not found: wdio-electron-service/preload

### DISCUSSION THREAD

**Discussion by mig8447** [webdriverio-community/wdio-electron-service#667] [Q&A] [answered]
Hello,

- I'm using electron with JS (Not TS)
- Specifically using `require` (CJS) for importing modules
- I could make the examples work by changing the extension of the test file to `mjs` so that it is treated as a module
- When trying to enable `browser.electron` I added the following line to `main.js`:
    ```js
    if ( process.env.TEST === 'true' ) require( 'wdio-electron-service/main' );
    ```
    
    and I added the following to my `preload.js`
    
    ```js
    if ( process.env.TEST === 'true' ) require( 'wdio-electron-service/preload' );
    ```
- The require for main works without an issue, however, the require for preload throws the following error in devtools:
    ```txt
    Error: module not found: wdio-electron-service/preload
        at preloadRequire (VM4 sandbox_bundle:2:85095)
        at <anonymous>:7:39
        at runPreloadScript (VM4 sandbox_bundle:2:85847)
        at VM4 sandbox_bundle:2:86106
        at VM4 sandbox_bundle:2:86261
        at ___electron_webpack_init__ (VM4 sandbox_bundle:2:86265)
        at VM4 sandbox_bundle:2:86388
    ```
- I found the following note at https://www.electronjs.org/docs/latest/tutorial/sandbox#preload-scripts

    > Because the require function is a polyfill with limited functionality, you will not be able to use [CommonJS modules](https://nodejs.org/api/modules.html#modules_modules_commonjs_modules) to separate your preload script into multiple files. If you need to split your preload code, use a bundler such as [webpack](https://webpack.js.org/) or [Parcel](https://parceljs.org/).
    
    Which I suspect is the cause for the module not being found

Would you know how to make this work using Vanilla CJS?

---

Update: 

- I added webpack and bundled the preload script without success
- I also disabled the sandbox without success.
- By now when the application is packaged/built with electron-builder, a message appears saying:

    ```txt
    Error: module not found: wdio-electron-service/main
    ```

I'm not sure what else to try

---

Update:

I bumped into https://github.com/goosewobbler/wdio-electron-service-example which is a perfect example of a testable CJS electron application. Hope it helps others too.

I found out that when loading a web extension into Electron, WDIO opens the context of the extension window, and if that window doesn't use the preload script then the test fails.

**Comment by mig8447** [comment, marked as answer]
I figured it out. I am still not sure why, I was going deeper into the matter (Dig into the chromedriver launching code) and I thought of "moving" the extension loading code to the end of the "ready" event handler and that did the trick.

So bottom line, **if you're loading extensions into Electron, you should do that at the end of the "ready" event handler. Otherwise, the chromedriver will attach to the first extension's background page.**

**Reply by goosewobbler** [reply]
This is useful info, I'll add it to the docs.
